### ISSUE TITLE
Bug: [strict-boolean-expressions] allowNullableBoolean - support array predicate functions

### ISSUE THREAD

**Issue by xkabylgSICKAG** [typescript-eslint/typescript-eslint#10688]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.7.2&fileType=.tsx&code=LAKAxg9gdgzgLgAgIYC4EAoCuUAmBTAMwEso8cEAfBAIwggBs8kpKEpN76BKAbQF0EAXgQ84AJ0x4ANAmz5ipHDPac%2BAblBIAdDAgBbPOnQAPLkIB8CUxpChIsRNTTpaDJiyoru-ISPGTlDnp1UGodfUMTM0FLa1A7aHgEMGdXRmZWOUISMl4BYVEJaVlcbMUQ8HCDI1MLKy4bBIcEHFS6dKg830LJPlAcKsjamPqgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUFs6gAQBcBPABxQGNoBLEggWhXioDsCB6ZA68%2BgIwHt%2BSAIbMGADxLQUyKv2aoMAbXC4IkRNGj9okADSq1OYFGHx4-AO4A5BPGG8kAIUEjm6MF1iIwAX0NGALqq-r5AA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
const a: (undefined | boolean | null)[] = [true, undefined, null];
a.some((x) => x);
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
          "@typescript-eslint/strict-boolean-expressions": [
        "error",
            { "allowNullableBoolean": true }
          ]
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

I expected there to be no errors.

### Actual Result

It seems like the array functions doesn't respect the "allowNullableBoolean" option

### Additional Info

Probably introduced in this PR https://github.com/typescript-eslint/typescript-eslint/pull/10106

**Comment by fuadsaud** [comment]
This is useful for logic expressions that use null coalescing, e.g.:

```
arr.filter(x =>
  x.y > 42 && x?.z > 0
)
```

**Comment by kirkwaiblinger** [comment]
Just noting for conversation - this isn't a bug about the `allowNullableBoolean` specifically. It's just that the array predicate method check doesn't respect the options at all (and always demands a `boolean` from array predicates). It probably does make sense to respect the options (though `boolean` really is better in these cases since it makes predicates eligible for inferred type predicates!)

**Comment by fuadsaud** [comment]
Thanks, @kirkwaiblinger, that makes sense!

Could you clarify what you mean by

> though boolean really is better in these cases since it makes predicates eligible for inferred type predicates!

I'm not sure I follow.

**Comment by kirkwaiblinger** [comment]
> Could you clarify what you mean by
> 
> > though boolean really is better in these cases since it makes predicates eligible for inferred type predicates!
> 
> I'm not sure I follow.

Oh, yeah, basically I'm just saying:

([playground](https://typescript-eslint.io/play/#ts=5.7.2&fileType=.tsx&code=MYewdgzgLgBAhgJwXAngeQGYDkCuAbPOAIzwFMIYBeGAbTHzwBoYAiMEWeglgXQG4AUANCRO4XAWJkIAITgATKvCSpMEwiXIA6DAEs8UUggAUADyoA%2BGKYCUggPT2YAFQDKMAKIAlL2i8wAWmsYXQoABxAICF0SFBguPAF2MHUpcjl5HRAEDzhgAAszSxgRCBAyLTwQAHMzLSgQABkQAHcjAGE4CFJjGz7BYXBoePEGNIgAcRAQRWpEZHRsMc0IHX1DE3NKK3MAQmoEuwFHGBbsgGsIXaTRyRWpmaycvMKtq1Ly0kqauobmtoQnW6vX6QA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false))
```ts
const arrayOfNullables = [null, "not null"];

const nonNullablesBad = arrayOfNullables.filter(x => x);
// TS ERROR - x is possibly null
nonNullablesBad.forEach(x => console.log(x.toLowerCase()));

const nonNullablesGood = arrayOfNullables.filter(x => x != null);
// works!
nonNullablesGood.forEach(x => console.log(x.toLowerCase()));
```

For further reading on inferred type predicates, see [the TS 5.5 release notes](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-5.html). Also related, https://github.com/typescript-eslint/typescript-eslint/issues/9764

***

So, IMO there's additional benefit in TS to using an actual boolean return in array predicate callbacks compared to the ordinary stylistic reasons for using strict-boolean-expressions.

**Comment by fuadsaud** [comment]
Thanks! I was not familiar with inferred type predicates, makes sense.
